### PR TITLE
WT-3474 Extend Python timestamp rollback test

### PR DESCRIPTION
**Pull request by bvpvamsikrishna**
@agorrod  @sueloverso  @michaelcahill 

initial commit to share Hang testcase.
When we have log enabled on both connection and table, then the call to rollback_to_stable is hanged.

When we run test_timestamp04.py from this pull request, the system is hung.
 
You can comment "'nolog'" in conncfg option to save 4 seconds before seeing hang.
Request to check

This pull request also includes the correction in error msg when the lsm cache size configured is smaller than required.

### CODE REVIEW

**Comment by michaelcahill** [comment]
autobuild: retest this please

**Comment by michaelcahill** [comment]
@bvpvamsikrishna, I pushed a small change to use `assertTrue(actual == expected)` rather than `assertEqual(actual, expected)`.  When there is a mismatch with large objects, `assertEqual` is trying to describe the differences and that process is slow (and appears to be recursive, leading to stack overflows).

**Comment by sueloverso** [comment]
The PR tester is failing on the assertTrue statements.  But also this change only partly covers the test cases in the ticket.  The changes now test logging enabled/disabled on the entire connection.  But there is not yet a test case for having logging enabled/disabled on the table itself which is how MongoDB will use it.

**Comment by sueloverso** [comment]
@bvpvamsikrishna per @michaelcahill's change for `assertTrue` versus `assertEqual` can you check/fix the other timestamp tests for that?  I know `test_timestamp03.py` and `test_timestamp07.py` at a minimum also use that construct, possibly others.

**Comment by bvpvamsikrishna** [comment]
@sueloverso for test_timestamp03.py and test_timestamp07.py also, I will change to use assertTrue for assertEqual.
@michaelcahill 
Thank you for change.

**Comment by bvpvamsikrishna** [comment]
pending work:
logged connection cases

**Comment by bvpvamsikrishna** [comment]
@sulabhM @nehakhatri5 
Request to review the pull request.

File: src/lsm/lsm_tree.c [hunk 1]
```diff
@@ -455,7 +455,7 @@ __lsm_tree_open_check(WT_SESSION_IMPL *session, WT_LSM_TREE *lsm_tree)
 		    "must be at least %" PRIu64 " (%" PRIu64 "MB)",
 		    S2C(session)->cache_size,
 		    S2C(session)->cache_size / WT_MEGABYTE,
-		    required, required / WT_MEGABYTE);
+		    required, (required + (WT_MEGABYTE - 1))/ WT_MEGABYTE);
```
Comment (sulabhM) [inline src/lsm/lsm_tree.c]: math.h offers a ceil() function. Can we do either of the two:
1. include math.h and use ceil() ( *I am not sure if we want to add a header file just to use this* )
2. Write a CEIL(x) macro to round up x to nearest integer >= x

File: test/suite/test_timestamp04.py [hunk 2]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+    #]
+
+    #types = [
+        #('col_fix', dict(empty=1, uri='table:', extra_config=',key_format=r, value_format=8t',is_lsm=False)),
+        #('col_var', dict(empty=0, uri='table:', extra_config=',key_format=r',is_lsm=False)),
+        #('lsm', dict(empty=0, uri='lsm:', extra_config=',type=lsm',is_lsm=True)),
+        #('row', dict(empty=0, uri='table:', extra_config='',is_lsm=False)),
+    #]
 
     scenarios = make_scenarios([
```
Comment (sulabhM) [inline test/suite/test_timestamp04.py]: test_compact02.py seems to utilise overriding standard connection open methods to create connections for several scenarios based on different cache sizes. Can we try something like that here in this test.

File: test/suite/test_timestamp04.py [hunk 3]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+            self.table_ts_nolog, dict((k, (2 if j <= (key_range / 4) else 1))
             for j, k in enumerate(keys[:(key_range / 2)])))
-        self.check(self.session, 'read_timestamp=' + timestamp_str(key_range / 2),
-            dict((k, 1) for k in keys[(1 + key_range / 2):]), missing=True)
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_ts_nolog, dict((k, 1) for k in keys[(1 + key_range / 2):]), missing=True)
+
+        # behavior of logging tables changes for rollback
+        if self.using_log == True:
+            # if log is enabled, none of the keys will be rolledback.
```
Comment (sulabhM) [inline test/suite/test_timestamp04.py]: "rolled back"

File: test/suite/test_timestamp04.py [hunk 4]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+            # Check that we see all the keys
+            self.check(self.session, 'read_timestamp=' + latest_ts,
+                self.table_ts_log, dict((k, 2) for k in keys[:]))
+        else:
+            # if log is disabled, keys will be rolledback till stable_timestamp
```
Comment (sulabhM) [inline test/suite/test_timestamp04.py]: rolled back

File: test/suite/test_timestamp04.py [hunk 5]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_ts_nolog, dict((k, 1) for k in keys[(1 + key_range / 2):]), missing=True)
+
+        # behavior of logging tables changes for rollback
+        if self.using_log == True:
+            # if log is enabled, none of the keys will be rolledback.
+            # Check that we see all the keys
+            self.check(self.session, 'read_timestamp=' + latest_ts,
+                self.table_ts_log, dict((k, 2) for k in keys[:]))
+        else:
```
Comment (nehakhatri5) [inline test/suite/test_timestamp04.py]: Aren't these tests same as the tests starting at line # 262

File: src/lsm/lsm_tree.c (hunk 1 above)
Comment (bvpvamsikrishna) [reply src/lsm/lsm_tree.c]: To use ceil is making code complex, as it is flagging bad-function-cast error as ceil returns double and we expect unsigned long long.

File: test/suite/test_timestamp04.py (hunk 2 above)
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 6]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
-    # Rollback only works for non-durable tables
-    conn_config = 'cache_size=20MB,log=(enabled=false)'
-
     # Check that a cursor (optionally started in a new transaction), sees the
     # expected values.
-    def check(self, session, txn_config, expected, missing=False):
+    def check(self, session, txn_config, tablename, expected, missing=False, prn=False):
         if txn_config:
             session.begin_transaction(txn_config)
-        c = session.open_cursor(self.uri, None)
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: cur is more readable, else sometimes confused what c represents, trying to give more readable names to variables.

File: test/suite/test_timestamp04.py [hunk 7]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        cur = session.open_cursor(self.uri + tablename, None)
         if missing == False:
-            actual = dict((k, v) for k, v in c if v != 0)
-            #print expected
-            #print actual
-            self.assertEqual(actual, expected)
+            actual = dict((k, v) for k, v in cur if v != 0)
+            if prn == True:
+                print "CHECK : Expected "
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 8]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+                print expected
+                print "CHECK : Actual   "
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 9]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_nots_log, dict((k, 1) for k in keys[:]))
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_nots_nolog, dict((k, 1) for k in keys[:]))
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_ts_log, dict((k, 1) for k in keys[:]))
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_ts_nolog, dict((k, 1) for k in keys[:]))
+
+        #Scenario: 2
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 10]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
-        self.check(self.session, 'read_timestamp=' + timestamp_str(key_range / 2),
-            dict((k, 1) for k in keys[(key_range / 2 + 1):]), missing=True)
+        # Check that we see the inserted value (i.e. 1) for all the keys in
+        # non-timestamp tables
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_nots_log, dict((k, 1) for k in keys[:]))
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_nots_nolog, dict((k, 1) for k in keys[:]))
+
+        # if nologging tables the behavior is consistent across connections
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 11]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        # with or without log enabled
+        # Check that we see the inserted value (i.e. 1) for the keys in a
+        # timestamp table till the stable_timestamp only.
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_ts_nolog, dict((k, 1) for k in keys[:(key_range / 2)]))
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_ts_nolog, dict((k, 1) for k in keys[(key_range / 2 + 1):]), missing=True)
+
+        # behavior of logging tables changes for rollback
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 12]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        if self.using_log == True:
+            # if log is enabled, none of the keys will be rolledback.
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 13]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+            # Check that we see all the keys
+            self.check(self.session, 'read_timestamp=' + latest_ts,
+                self.table_ts_log, dict((k, 1) for k in keys[:]))
+        else:
+            # if log is disabled, keys will be rolledback till stable_timestamp
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 14]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_nots_nolog, dict((k, 2) for k in keys[:]))
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_ts_log, dict((k, 2) for k in keys[:]))
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_ts_nolog, dict((k, 2) for k in keys[:]))
 
-        # Rollback a quater of the new commits
-        self.conn.set_timestamp('stable_timestamp=' + timestamp_str(1 + key_range + key_range / 4))
+        #Scenario: 4
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 15]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        # Advance the stable_timestamp by a quater range and rollback.
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 16]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        # three-quarter timestamps will be rolledback.
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 17]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
-        # second half have a value of 1
-        self.check(self.session, 'read_timestamp=' + timestamp_str(2 * key_range),
-            dict((k, (2 if j <= (key_range / 4) else 1))
+        # Check that we see the updated value (i.e. 2) for all the keys in
+        # non-timestamp tables
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_nots_log, dict((k, 2) for k in keys[:]))
+        self.check(self.session, 'read_timestamp=' + latest_ts,
+            self.table_nots_nolog, dict((k, 2) for k in keys[:]))
+        # if nologging tables the behavior is consistent across connections
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py (hunk 3 above)
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py (hunk 4 above)
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done.

File: test/suite/test_timestamp04.py [hunk 18]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
+        cur.close()
         if txn_config:
             session.commit_transaction()
 
-    def test_basic(self):
+    def test_rollback_to_stable(self):
         if not wiredtiger.timestamp_build():
             self.skipTest('requires a timestamp build')
 
+        uri_ts_log      = self.uri + self.table_ts_log
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: There is nothing like partially logged tables, clarified with alex.

File: test/suite/test_timestamp04.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: Done

File: test/suite/test_timestamp04.py [hunk 19]
```diff
@@ -44,103 +44,244 @@ def timestamp_ret_str(t):
...
-            c[k] = 2
+            cur_ts_log[k] = 2
+            cur_ts_nolog[k] = 2
             self.session.commit_transaction('commit_timestamp=' + timestamp_str(k + key_range))
 
-        # Now we should have: keys 1-100 with value 2
-        self.check(self.session, 'read_timestamp=' + timestamp_str(2 * key_range),
-            dict((k, 2) for k in keys[:]))
+        # Scenario: 3
+        # Check that we see all values updated (i.e 2) in all tables
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: yes, for better maintainability

File: test/suite/test_timestamp04.py (hunk 5 above)
Comment (bvpvamsikrishna) [reply test/suite/test_timestamp04.py]: initial tests are for inserts, the second round covers updates aswell.

**Comment by bvpvamsikrishna** [comment]
autobuild : retest this please

**Comment by bvpvamsikrishna** [comment]
@sulabhM @nehakhatri5 
Request to check updates for comments.

File: src/lsm/lsm_tree.c (hunk 1 above)
Comment (sulabhM) [reply src/lsm/lsm_tree.c]: Can we get a macro instead then, maybe something like: #define CEIL(X, Y) ((X) + ((Y) - 1) / (Y))

**Comment by bvpvamsikrishna** [comment]
@agorrod 
Request to review this pull request.

**Comment by agorrod** [comment]
Thanks @bvpvamsikrishna lgtm

**Pull request merged**